majority-multisign/src/MajorityMultiSign/Schema.hs

```diff
@@ -12,67 +12,83 @@
 
 module MajorityMultiSign.Schema (
   MajorityMultiSign,
-  MajorityMultiSignDatum (..),
-  MajorityMultiSignIdentifier (..),
-  MajorityMultiSignRedeemer (..),
+  MajorityMultiSignDatum (MajorityMultiSignDatum, signers),
+  MajorityMultiSignIdentifier (MajorityMultiSignIdentifier, asset),
+  MajorityMultiSignRedeemer (UpdateKeysAct, UseSignaturesAct),
   MajorityMultiSignSchema,
-  MajorityMultiSignValidatorParams (..),
-  SetSignaturesParams (..),
+  MajorityMultiSignValidatorParams (MajorityMultiSignValidatorParams, asset),
+  SetSignaturesParams (SetSignaturesParams, mmsIdentifier, newKeys, pubKeys),
   getMinSigners,
+  maximumSigners,
+  naturalLength,
 ) where
 
-import Data.Aeson (FromJSON, ToJSON)
+import Data.Aeson (FromJSON, ToJSON, (.:), (.=))
+import Data.Aeson qualified as Aeson
 import Data.Aeson.Extras (encodeByteString)
-import Data.Functor.Foldable (Fix (..))
+import Data.Functor.Foldable (Fix (Fix))
 import Data.OpenApi.Schema qualified as OpenApi
+import Data.Semigroup (Sum (Sum, getSum))
 import Data.Text qualified as Text
 import Data.Text.Encoding (decodeUtf8)
 import GHC.Generics (Generic)
 import Ledger.Crypto (PubKey)
 import Ledger.Typed.Scripts qualified as Scripts
 import Plutus.Contract (Endpoint, type (.\/))
 import Plutus.V1.Ledger.Api (PubKeyHash)
-import Plutus.V1.Ledger.Value (AssetClass, CurrencySymbol (..), TokenName (..))
+import Plutus.V1.Ledger.Value (AssetClass, CurrencySymbol (CurrencySymbol), TokenName (TokenName))
 import PlutusTx qualified
 import PlutusTx.NatRatio (NatRatio, ceiling, frac, fromNatural)
-import PlutusTx.Natural (Natural)
-import PlutusTx.Prelude hiding (Eq, decodeUtf8)
+import PlutusTx.Natural (Natural, nat)
+import PlutusTx.Prelude hiding (Eq, decodeUtf8, (<$>), (<*>))
 import Schema (
-  FormArgumentF (..),
-  FormSchema (..),
+  FormArgumentF (FormHexF, FormMaybeF, FormObjectF, FormStringF),
+  FormSchema (FormSchemaMaybe, FormSchemaObject),
   ToArgument (toArgument),
   ToSchema (toSchema),
  )
-import Prelude (Eq, Show)
+import Prelude (Eq, Show, (<$>), (<*>))
 
 {-# INLINEABLE signReq #-}
 
 -- | Signing proportion required
 signReq :: NatRatio
 signReq = [frac| (1, 2) |] -- 0.5
 
-{-# INLINEABLE intToNatRatio #-}
-intToNatRatio :: Integer -> NatRatio
-intToNatRatio = fromNatural . toEnum @Natural
+{-# INLINEABLE maximumSigners #-}
 
-{-# INLINEABLE ceilNatRatioToInt #-}
-ceilNatRatioToInt :: NatRatio -> Integer
-ceilNatRatioToInt = fromEnum . ceiling
+-- | Maximum number of signers allowed
+maximumSigners :: Natural
+maximumSigners = [nat| 10 |]
+
+{-# INLINEABLE naturalLength #-}
+
+-- | A count of the items in a `Foldable` is always 'Natural'.
+naturalLength :: Foldable t => t a -> Natural
+naturalLength = getSum . foldMap (Sum . const [nat| 1 |])
```

Comment:
Will starting at 1 here not mean the length is always 1 too high?
Also, could we do this with just fold and succ?
`naturalLength = foldr (const succ) zero`

Reply:
> Will starting at 1 here not mean the length is always 1 too high?

It would be if it was the first argument of `foldr` rather than `foldMap`. This is replacing every item of the `Foldable` with `Sum 1` and then folding them.

> Also, could we do this with just fold and succ? `naturalLength = foldr (const succ) zero`

Yes, but Koz doesn't like `Enum` and when I tried something similar it turned out to be longer (code size wise): https://github.com/Liqwid-Labs/plutus-extra/issues/81

Reply:
Ah alright, makes sense